File: app/src/main/java/com/prey/actions/triggers/TriggerParse.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2019 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.triggers;

import android.content.Context;

import com.prey.PreyLogger;
import com.prey.net.PreyWebServices;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class TriggerParse {

    public static List<TriggerDto> getJSONFromUrl(Context ctx) {
        String json = null;
        try {
            json = PreyWebServices.getInstance().triggers(ctx);
            return getJSONFromTxt(ctx, json);
        } catch (Exception e) {
            return null;
        }
    }

    public static List<TriggerDto> getJSONFromTxt(Context ctx, String json) {
        json = "{\"prey\":" + json + "}";
        List<TriggerDto> listTrigger = new ArrayList<TriggerDto>();
        PreyLogger.d(json);
        try {
            JSONObject jsnobject = new JSONObject(json);
            JSONArray jsonArray = jsnobject.getJSONArray("prey");
            for (int i = 0; i < jsonArray.length(); i++) {
                String jsonCommand = jsonArray.get(i).toString();
                JSONObject explrObject = new JSONObject(jsonCommand);
                TriggerDto trigger = new TriggerDto();
                try {
                    trigger.id = explrObject.getInt("id");
                } catch (Exception e) {
                    trigger.id = 101;
                }
                try {
                    trigger.name = explrObject.getString("name");
                } catch (Exception e) {
                    trigger.name = "ups";
                }
                trigger.events = explrObject.getString("automation_events");
                trigger.actions = explrObject.getString("automation_actions");
                listTrigger.add(trigger);
            }
        } catch (Exception e) {
            PreyLogger.e("e:" + e.getMessage(), e);
            return null;
        }
        return listTrigger;
    }

    public static List<TriggerEventDto> TriggerEvents(String events) {
        events = "{\"prey\":" + events + "}";
        List<TriggerEventDto> listTrigger = new ArrayList<TriggerEventDto>();
        PreyLogger.d(events);
        try {
            JSONObject jsnobject = new JSONObject(events);
            JSONArray jsonArray = jsnobject.getJSONArray("prey");
            for (int i = 0; i < jsonArray.length(); i++) {
                String jsonCommand = jsonArray.get(i).toString();
                JSONObject explrObject = new JSONObject(jsonCommand);
                TriggerEventDto trigger = new TriggerEventDto();
                trigger.type = explrObject.getString("type");
                trigger.info = explrObject.getString("info");
                listTrigger.add(trigger);
            }
        } catch (Exception e) {
            return null;
        }
        return listTrigger;
    }

    public static List<TriggerActionDto> TriggerActions(String actions) {
        actions = "{\"prey\":" + actions + "}";
        List<TriggerActionDto> listTrigger = new ArrayList<TriggerActionDto>();
        PreyLogger.d(actions);
        try {
            JSONObject jsnobject = new JSONObject(actions);
            JSONArray jsonArray = jsnobject.getJSONArray("prey");
            for (int i = 0; i < jsonArray.length(); i++) {
                String jsonCommand = jsonArray.get(i).toString();
                JSONObject explrObject = new JSONObject(jsonCommand);
                TriggerActionDto trigger = new TriggerActionDto();
                trigger.delay = explrObject.getInt("delay");
                trigger.action = explrObject.getString("action");
                listTrigger.add(trigger);
            }
        } catch (Exception e) {
            return null;
        }
        return listTrigger;
    }

}

File: app/src/main/java/com/prey/actions/triggers/TriggerParse.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.triggers

import android.content.Context
import com.prey.PreyConfig

import com.prey.PreyLogger
import com.prey.net.PreyWebServices

import org.json.JSONObject

/**
 * TriggerParse is an object that provides methods for parsing trigger data from JSON.
 */
object TriggerParse {

    /**
     * Retrieves trigger data from a URL and parses it into a list of TriggerDto objects.
     *
     * @param context The application context.
     * @return A list of TriggerDto objects, or null if an error occurs.
     */
    fun getJSONFromUrl(context: Context): List<TriggerDto>? {
        var json: String? = null
        try {
            json = PreyConfig.getInstance(context).getWebServices().triggers(context)
            if (json != null) {
                return getJSONFromTxt(context, json)
            }
        } catch (e: Exception) {
            PreyLogger.e("Error:${e.message}", e)

        }
        return null
    }

    /**
     * Parses a JSON string into a list of TriggerDto objects.
     *
     * @param context The application context.
     * @param jsonValue The JSON string to parse.
     * @return A list of TriggerDto objects, or null if an error occurs.
     */
    fun getJSONFromTxt(context: Context, jsonValue: String): List<TriggerDto>? {
        val listTrigger: MutableList<TriggerDto> = ArrayList()
        if (jsonValue != null && "" != jsonValue) {
            var json = "{\"prey\":${jsonValue}}"
            PreyLogger.d(json)
            try {
                val jsnobject = JSONObject(json)
                val jsonArray = jsnobject.getJSONArray("prey")
                for (i in 0 until jsonArray.length()) {
                    val jsonCommand = jsonArray[i].toString()
                    val explrObject = JSONObject(jsonCommand)
                    val trigger = TriggerDto()
                    try {
                        trigger.setId(explrObject.getInt("id"))
                    } catch (e: java.lang.Exception) {
                        trigger.setId(101)
                    }
                    try {
                        trigger.setName(explrObject.getString("name"))
                    } catch (e: java.lang.Exception) {
                        trigger.setName("ups")
                    }
                    trigger.setEvents(explrObject.getString("automation_events"))
                    trigger.setActions(explrObject.getString("automation_actions"))
                    listTrigger.add(trigger)
                }
            } catch (e: java.lang.Exception) {
                PreyLogger.e("Error: ${e.message}", e)
                return null
            }
        }
        return listTrigger
    }

    /**
     * Parses a JSON string into a list of TriggerEventDto objects.
     *
     * @param events The JSON string to parse.
     * @return A list of TriggerEventDto objects, or null if an error occurs.
     */
    fun TriggerEvents(events: String): List<TriggerEventDto>? {
        var events = events
        events = "{\"prey\":$events}"
        val listTrigger: MutableList<TriggerEventDto> = ArrayList()
        PreyLogger.d(events)
        try {
            val jsnobject = JSONObject(events)
            val jsonArray = jsnobject.getJSONArray("prey")
            for (i in 0 until jsonArray.length()) {
                val jsonCommand = jsonArray[i].toString()
                val explrObject = JSONObject(jsonCommand)
                val trigger = TriggerEventDto()
                trigger.setType(explrObject.getString("type"))
                trigger.setInfo(explrObject.getString("info"))
                listTrigger.add(trigger)
            }
        } catch (e: Exception) {
            return null
        }
        return listTrigger
    }

    /**
     * Parses a JSON string into a list of TriggerActionDto objects.
     *
     * @param actions The JSON string to parse.
     * @return A list of TriggerActionDto objects, or null if an error occurs.
     */
    fun TriggerActions(actions: String): List<TriggerActionDto>? {
        var actions = actions
        actions = "{\"prey\":$actions}"
        val listTrigger: MutableList<TriggerActionDto> = ArrayList()
        PreyLogger.d(actions)
        try {
            val jsnobject = JSONObject(actions)
            val jsonArray = jsnobject.getJSONArray("prey")
            for (i in 0 until jsonArray.length()) {
                val jsonCommand = jsonArray[i].toString()
                val explrObject = JSONObject(jsonCommand)
                val trigger = TriggerActionDto()
                trigger.setDelay(explrObject.getInt("delay"))
                trigger.setAction(explrObject.getString("action"))
                listTrigger.add(trigger)
            }
        } catch (e: Exception) {
            return null
        }
        return listTrigger
    }

}